python/tvm/relay/op/nn/_nn.py

```diff
@@ -125,20 +125,22 @@ def schedule_conv2d(attrs, outs, target):
     groups = attrs.groups
     layout = attrs.data_layout
     kernel_layout = attrs.kernel_layout
+    channels = attrs.channels #TODO should be input channels
```

Comment:
it is not safe to use `attrs.channels` since this attr is optinoal

Reply:
In that case, the `in_channels` needs to be exposed to the function.  Either by being added to the `attrs` (defined at [nn.h:51](https://github.com/dmlc/tvm/blob/8d3b392da6eae59dab068947e76ddf9e3534661d/include/tvm/relay/attrs/nn.h#L51)), or passing it as an extra argument (such as the `tinfo` [suggestion](https://github.com/dmlc/tvm/pull/3020#issuecomment-482809280)) by every caller of it in the codebase.  

Would there ever be any cases when the `in_channels` would be indeterminate when this function is called?

Reply:
`in_channels` should be available from the shape of input / weight after type infer

Reply:
@Wheest As suggested in https://github.com/dmlc/tvm/pull/3070#issuecomment-485536853, you can get input channels via `outs[0].op.attrs["workload"]`. Can you check if this work for you?

Reply:
Thanks for the insight @vinx13, I would never have groked all of the info that `outs` stores.  I can confirm that setting `in_channels = outs[0].op.attrs['workload'][1][1]` works for `groups==1` and `groups==in_channels`.  However, interestingly when `1<groups<in_channels`, it seems that `outs[0].op.attrs.items` is empty.  This suggests that whatever calls `schedule_conv2d` does not set `outs[0].op.attrs.items` correctly when using grouped convolutions.  Am investigating why.

Reply:
Hi @Wheest, I did notice such circumstances in my PR. Please let me know if you have any new discovery :)

Reply:
@Wheest FYI `workload` is set [here](https://github.com/dmlc/tvm/blob/3f835bd974536a3b19dc63d7f7e55ada01d752bd/python/tvm/autotvm/task/topi_integration.py#L333)
This is probably because `group_conv2d_nchw` is not registered as autotvm template

Reply:
@vinx13 You might be correct - `depthwise_conv2d_nchw` has been properly registered while `group_conv2d_nchw` has not. I'm working on fixing that

Reply:
Thanks, I'll hold until then.  Why would autotvm templates be used in this workflow?  I thought that the process is:
1. Define model in using high level Relay API (with the bug in this PR happening here).
2. Generate model description in Relay IR
3. Compile model to chosen backend.
As I understand it, you might use autotvm at step 3. to tune and optimise the program.  But I thought of it a "separate tool" in the tvm suite, not used in step 1 or 2.

Reply:
Actually we only need extra attributes when calling schedule. One way is to add attributes needed in topi compute. Autotvm already did this, but we can also register them manually. For the group conv case, we can add an attribute when calling `tvm.compute`